### PR TITLE
feat: Introduce `enforce-hash-update` flag to help hashing migration

### PR DESCRIPTION
**Pull request by pgrzesik**
Reported internally, introduce a way to ease the lambda hashing algorithm migration

### CODE REVIEW

File: docs/guides/upgrading-v3.md [hunk 1]
```diff
@@ -196,6 +196,30 @@ That allowed us to make the KMS configuration consistent with all other AWS reso
 
 `alexaSkill` events now require an `appId` ([learn more](../deprecations.md#support-for-alexaskill-event-without-appid-is-to-be-removed)). That change was required to implement a more stable deployment, as well as to deploy more restricted IAM permissions.
 
+### Lambda Hashing Algorithm
+
+By default, Lambda version hashes will now be generated using improved algorithm (fixed determinism issues). As it is a breaking change that requires more manual effort during migration, it is still possible (but not recommended) to keep using old algorithm by using the following configuration:
+
+````
```
Comment (medikoo) [inline docs/guides/upgrading-v3.md:203]: I think it should be one back quote less

File: docs/guides/upgrading-v3.md [hunk 2]
```diff
@@ -236,7 +260,7 @@ provider:
       statements: ...
       managedPolicies: ...
     deploymentRole: ...
-```
+````
```
Comment (medikoo) [inline docs/guides/upgrading-v3.md:263]: And then this change should not happen

**Review by medikoo — COMMENTED** [review]
Well done! Looks as a great migration aid

I've just spotted one styling issue in readme

**Review by medikoo — APPROVED** [review]
Looks great 👍

File: docs/guides/upgrading-v3.md [hunk 3]
```diff
@@ -196,6 +196,31 @@ That allowed us to make the KMS configuration consistent with all other AWS reso
 
 `alexaSkill` events now require an `appId` ([learn more](../deprecations.md#support-for-alexaskill-event-without-appid-is-to-be-removed)). That change was required to implement a more stable deployment, as well as to deploy more restricted IAM permissions.
 
+### Lambda Hashing Algorithm
+
+By default, Lambda version hashes will now be generated using improved algorithm (fixed determinism issues). As it is a breaking change that requires more manual effort during migration, it is still possible (but not recommended) to keep using old algorithm by using the following configuration:
```
Comment (mnapoli) [inline docs/guides/upgrading-v3.md:201]:
```suggestion
By default, Lambda version hashes will now be generated using an improved algorithm (fixes determinism issues). As it is a breaking change that requires more manual effort during migration, it is still possible (but not recommended) to keep using the old algorithm by using the following configuration:
```

File: docs/guides/upgrading-v3.md [hunk 4]
```diff
@@ -196,6 +196,31 @@ That allowed us to make the KMS configuration consistent with all other AWS reso
...
+
+```
+provider:
+  lambdaHashingVersion: 20200924
+```
+
+However, we highly encourage an upgrade to the new algorithm. Safe upgrade requires change in configuration or code of all your functions. You can do it safely following the guide below:
```
Comment (mnapoli) [inline docs/guides/upgrading-v3.md:208]: "Safe upgrade" makes me think (as a user) that something may not be safe (loosing stuff? destroying resources accidentally?)

I'd rephrase with something like:

```suggestion
However, we highly encourage an upgrade to the new algorithm. To upgrade, you must redeploy your service with code or configuration changes in all functions. You can do it by following the guide below:
```

File: docs/guides/upgrading-v3.md [hunk 5]
```diff
@@ -196,6 +196,31 @@ That allowed us to make the KMS configuration consistent with all other AWS reso
...
+
+**NOTE**: Please keep in mind that these changes require two deployments with manual configuration adjustment between them. It also creates two additional versions and temporarily overrides descriptions of your functions. Migration will need to be done separately for each of your environments/stages.
+
+1. Run `sls deploy` with additional `--enforce-hash-update` flag to override description for all your functions and force creation of new versions.
```
Comment (mnapoli) [inline docs/guides/upgrading-v3.md:212]:
```suggestion
1. Run `sls deploy` with additional `--enforce-hash-update` flag: that flag will override the description for all Lambda functions, which will force the creation of new versions.
```

I would also state clearly that we start from a Serverless v2 application. Maybe it's a bit over the top to explicitly write that? But at least as a user it's 100% clear. Something like this:

```
1. In a Serverless Framework v2 project
```

(or maybe put that part in the paragraph above?)

File: docs/guides/upgrading-v3.md [hunk 6]
```diff
@@ -196,6 +196,31 @@ That allowed us to make the KMS configuration consistent with all other AWS reso
...
+2. Set `provider.lambdaHashingVersion` to `20201221` in your configuration.
```
Comment (mnapoli) [inline docs/guides/upgrading-v3.md:213]:
```suggestion
2. Set `provider.lambdaHashingVersion` to `20201221` in your configuration: your service will now always deploy with the new Lambda version hashes (which is the new default in v3).
```

(just so that users understand what this step does in regards to v2 and v3)

File: docs/guides/upgrading-v3.md [hunk 7]
```diff
@@ -196,6 +196,31 @@ That allowed us to make the KMS configuration consistent with all other AWS reso
...
+3. Run `sls deploy`, this time without additional `--enforce-hash-update` flag.
```
Comment (mnapoli) [inline docs/guides/upgrading-v3.md:214]:
```suggestion
3. Run `sls deploy`, this time without additional `--enforce-hash-update` flag: that will restore the original descriptions on all Lambda functions.
```

That will reassure users that things go back to normal.

File: docs/guides/upgrading-v3.md [hunk 8]
```diff
@@ -196,6 +196,31 @@ That allowed us to make the KMS configuration consistent with all other AWS reso
...
+
+Now your whole service will be fully migrated to new Lambda Hashing Algorithm.
```
Comment (mnapoli) [inline docs/guides/upgrading-v3.md:216]:
```suggestion
Now your whole service is fully migrated to the new Lambda Hashing Algorithm.
```
(the present tense was used in the sentences above)

**Review by mnapoli — COMMENTED** [review]
🎉  great!

I've used grammarly.com to add a few missing words (minor), and some suggestions to clarify a few parts.

File: docs/guides/upgrading-v3.md (hunk 5 above)
Comment (pgrzesik) [reply docs/guides/upgrading-v3.md:212]: I think we can add it, but I think it's also a bit redundant since this guide is about migrating from v2 to v3 specifically. Where would you exactly see a good place to add such information?

Comment (mnapoli) [reply docs/guides/upgrading-v3.md:212]: 👍  no problem then ignore that part

**Review by medikoo — APPROVED** [review]
👍

**Review by mnapoli — APPROVED** [review]
